HoloJS/HoloJsHost/RenderingContext2D.cpp

```diff
@@ -50,62 +54,283 @@ void RenderingContext2D::drawImage(HologramJS::API::ImageElement* imageElement,
+        m_figurePresent = false;
+    }
+
+    float3x2 tmp = m_session->Transform;
+    m_session->Transform = float3x2::identity();
+
+    Geometry::CanvasGeometry ^ geometry = Geometry::CanvasGeometry::CreatePath(m_pathBuilder);
+
+    auto brush = ref new CanvasLinearGradientBrush(m_canvasRenderTarget, stops);
+    brush->StartPoint = start;
+    brush->EndPoint = end;
+    m_session->FillGeometry(geometry, brush);
+
+    m_session->Transform = tmp;
+
+    beginPath();
+    m_pathBuilder->AddGeometry(geometry);
+}
+
+void RenderingContext2D::stroke(Color& color)
+{
+    if (m_pathBuilder == nullptr) return;
+    if (m_figurePresent) {
+        m_pathBuilder->EndFigure(Geometry::CanvasFigureLoop::Open);
+        m_figurePresent = false;
+    }
+
+    float3x2 tmp = m_session->Transform;
+    m_session->Transform = float3x2::identity();
+
+    Geometry::CanvasGeometry ^ geometry = Geometry::CanvasGeometry::CreatePath(m_pathBuilder);
+
+    Geometry::CanvasStrokeStyle ^ style = ref new Geometry::CanvasStrokeStyle();
+    style->StartCap = capStyle;
+    style->EndCap = capStyle;
+    style->LineJoin = joinStyle;
+
+    Color cm = color;
+    cm.A *= m_globalOpacity;
+
+    m_session->DrawGeometry(geometry, cm, lineWidth * tmp.m11, style);
+
+    m_session->Transform = tmp;
+
+    beginPath();
+    m_pathBuilder->AddGeometry(geometry);
+}
+
+void RenderingContext2D::strokeGradient(float2& start, float2& end, Platform::Array<CanvasGradientStop> ^ stops)
+{
+    if (m_pathBuilder == nullptr) return;
+    if (m_figurePresent) {
+        m_pathBuilder->EndFigure(Geometry::CanvasFigureLoop::Open);
+        m_figurePresent = false;
+    }
+
+    float3x2 tmp = m_session->Transform;
+    m_session->Transform = float3x2::identity();
+
+    Geometry::CanvasGeometry ^ geometry = Geometry::CanvasGeometry::CreatePath(m_pathBuilder);
+
+    auto brush = ref new CanvasLinearGradientBrush(m_canvasRenderTarget, stops);
+    brush->StartPoint = start;
+    brush->EndPoint = end;
+    m_session->DrawGeometry(geometry, brush);
+
+    m_session->Transform = tmp;
+
+    beginPath();
+    m_pathBuilder->AddGeometry(geometry);
+}
+
+void RenderingContext2D::setLineStyle(float _lineWidth,
+                                      Geometry::CanvasCapStyle capMode,
+                                      Geometry::CanvasLineJoin joinMode)
+{
+    lineWidth = _lineWidth;
+    capStyle = capMode;
+    joinStyle = joinMode;
+}
+
+void RenderingContext2D::setTransform(double a, double b, double c, double d, double e, double f)
+{
+    m_session->Transform = float3x2(a, b, c, d, e, f);
+}
+
+void RenderingContext2D::setGlobalOpacity(float opacity) { m_globalOpacity = opacity; }
+
 Platform::Array<unsigned char> ^ RenderingContext2D::getImageData(Rect& rect, unsigned int* stride)
 {
+    m_session->Flush();
     if (!m_isOptimizedBitmap) {
         *stride = rect.Width * m_bpp;
         return m_canvasRenderTarget->GetPixelBytes(rect.X, rect.Y, rect.Width, rect.Height);
```

Comment:
Un-premultiplying should be done here. These pixels are used as source for textures which assume un-premultiplied. The alpha values are effectively applied twice.